### PR TITLE
More infrastructure to bring EOS desktop back to life

### PR DESCRIPTION
**Pull request by mariospr**
This is a list of commits that I are needed to bring the desktop back to life and that, most of them, will be needed later on to implement further changes that are at the moment living in my WIP branch here: https://github.com/endlessm/gnome-shell/commits/wip/mario/icongrid

The main reason I'm proposing this PR now is because I will be away until June 4th and this way it should be easier for others to either resume the work I have in my WIP branch (where I have a more functional desktop, but still in a very rudimentary state) or to work on other tasks that can also depend on some of the bits contained here (e.g. implementing the "Endless button" requires the commit that adds `workspaceMonitor.js, that's why I put it first in the llist).

The other reasons for me to propose this specific patch set and not to include other commits from my WIP branch is because the commits suggested here should be fairly safe to integrate in master, since the most visible change is the removal of the "Frequent / All" selector in the apps mode in the overview, and the fact that such view now loads the icons & folders from the Endless desktop, instead of the system ones.

@cosimoc: Please review. If you do it before this Monday there's a small chance I can look into addressing comments (no promises though), otherwise feel free to adjust the commits yourself as you see fit. Thanks!

https://phabricator.endlessm.com/T17230

### CODE REVIEW

File: js/ui/appDisplay.js [hunk 1]
```diff
@@ -1089,6 +1116,13 @@ const AppSearchProvider = new Lang.Class({
     createResultObject: function (resultMeta) {
         let app = this._appSys.lookup_app(resultMeta['id']);
         return new AppIcon(app);
+    },
+
+    activateResult: function(appId) {
```
Comment (cosimoc) [inline js/ui/appDisplay.js]: I think the idea here is that the existing `activate()` method on `AppIcon` will be called no? Not sure you need this method.

File: js/misc/util.js [hunk 2]
```diff
@@ -398,3 +425,84 @@ function ensureActorVisibleInScrollView(scrollView, actor) {
...
+        return default_string;
+    } catch (e) {
+        logError(e, 'error while parsing Chromium preferences');
+        return null;
+    }
+
+    let root = parser.get_root().get_object();
+
+    let searchProviderDataNode = root.get_member('default_search_provider_data');
+    if (!searchProviderDataNode || searchProviderDataNode.get_node_type() != Json.NodeType.OBJECT) {
```
Comment (cosimoc) [inline js/misc/util.js]: Remove curly braces for single-line if (many occurrences in this commit)

File: js/ui/workspaceMonitor.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    _getVisibleApps: function() {
+        let runningApps = this._appSystem.get_running();
+        return runningApps.filter(function(app) {
+            let windows = app.get_windows();
+            for (let window of windows) {
+                // We do not count transient windows because of an issue with Audacity
+                // where a transient window was always being counted as visible even
+                // though it was minimized
+                if (window.get_transient_for()) {
```
Comment (cosimoc) [inline js/ui/workspaceMonitor.js]: Remove curly braces for single-line if (many occurrences in this commit)

File: data/gnome-shell-theme.gresource.xml [hunk 4]
```diff
@@ -48,5 +48,27 @@
     <file>hot-corner-rtl-symbolic.svg</file>
     <file>mini-icon-active-indicator.png</file>
     <file>system-logout.png</file>
+
```
Comment (mariospr) [inline data/gnome-shell-theme.gresource.xml]: Actually, this commit shouldn't have made it to this PR, it's part of my WIP branch. I'll remove it

File: js/ui/backgroundMenu.js [hunk 5]
```diff
@@ -18,8 +19,20 @@ const BackgroundMenu = new Lang.Class({
 
         this.addSettingsAction(_("Change Background…"), 'gnome-background-panel.desktop');
         this.addMenuItem(new PopupMenu.PopupSeparatorMenuItem());
-        this.addSettingsAction(_("Display Settings"), 'gnome-display-panel.desktop');
-        this.addSettingsAction(_("Settings"), 'gnome-control-center.desktop');
+
+        this.addAction(_("Add App"), Lang.bind(this, function() {
```
Comment (mariospr) [inline js/ui/backgroundMenu.js]: I don't think that's a good idea. `addSettingsAction()` is meant to be used for panels of gnome control center, and I don't think it fits well for opening something else like GS. Unless I'm missing something obvious I think that using `addAction()`, either with or without the app activation context, is probably the right thing to do.

File: src/shell-app-system.c [hunk 6]
```diff
@@ -156,15 +215,99 @@ stale_app_remove_func (gpointer key,
...
+  ShellApp *app;
+
+  g_hash_table_iter_init (&iter, self->priv->id_to_app);
+  while (g_hash_table_iter_next (&iter, NULL, (gpointer) &app))
+    {
+      GDesktopAppInfo *app_info = NULL;
+
+      if (app_is_stale (app))
+        {
+          // App is stale, we remove it
```
Comment (mariospr) [inline src/shell-app-system.c]: Ok (not mine, though)

File: src/shell-app-system.c [hunk 7]
```diff
@@ -282,6 +443,39 @@ shell_app_system_lookup_heuristic_basename (ShellAppSystem *system,
...
+ * Returns: (transfer none): A #ShellApp for @alias
+ */
+ShellApp *
+shell_app_system_lookup_alias (ShellAppSystem *system,
+                               const char     *alias)
+{
+  ShellApp *result;
+  const char *id;
+
+  if (alias == NULL)
```
Comment (mariospr) [inline src/shell-app-system.c]: Makes  sense

File: src/shell-app-system.c [hunk 8]
```diff
@@ -364,20 +565,45 @@ _shell_app_system_notify_app_state_changed (ShellAppSystem *self,
 {
   ShellAppState state = shell_app_get_state (app);
 
+  gchar *app_address = g_strdup_printf ("%p", app);
+  GDesktopAppInfo *app_info = shell_app_get_app_info (app);
+  const gchar *app_info_id = NULL;
+  if (app_info != NULL) {
```
Comment (mariospr) [inline src/shell-app-system.c]: Ok

File: src/shell-app-system.c [hunk 9]
```diff
@@ -364,20 +565,45 @@ _shell_app_system_notify_app_state_changed (ShellAppSystem *self,
...
       g_hash_table_insert (self->priv->running_apps, g_object_ref (app), NULL);
+      g_hash_table_remove (self->priv->starting_apps, app);
       break;
     case SHELL_APP_STATE_STARTING:
+      g_hash_table_insert (self->priv->starting_apps, g_object_ref (app), NULL);
       break;
     case SHELL_APP_STATE_STOPPED:
       g_hash_table_remove (self->priv->running_apps, app);
+
+      if (g_hash_table_remove (self->priv->running_apps, app) && app_info_id != NULL)
```
Comment (mariospr) [inline src/shell-app-system.c]: Good catch. That's a bug we currently have too

File: configure.ac [hunk 10]
```diff
@@ -104,7 +104,8 @@ SHARED_PCS="gio-unix-2.0 >= $GIO_MIN_VERSION
             gobject-introspection-1.0 >= $GOBJECT_INTROSPECTION_MIN_VERSION
```
Comment (mariospr) [inline configure.ac]: Very true, sorry for the sloppy catch-it-all commit

File: js/ui/appDisplay.js (hunk 1 above)
Comment (mariospr) [inline js/ui/appDisplay.js]: I took that from the older codebase, where there was no AppIcon being created. I'll remove this method

File: src/st/st-entry.c [hunk 11]
```diff
@@ -1172,6 +1384,8 @@ _st_entry_set_icon (StEntry       *entry,
 
       clutter_actor_set_reactive (*icon, TRUE);
       clutter_actor_add_child (CLUTTER_ACTOR (entry), *icon);
+      g_signal_connect (*icon, "button-press-event",
+                        G_CALLBACK (_st_entry_icon_press_cb), entry);
       g_signal_connect (*icon, "button-release-event",
                         G_CALLBACK (_st_entry_icon_press_cb), entry);
```
Comment (mariospr) [inline src/st/st-entry.c]: Indeed

File: src/st/st-entry.c [hunk 12]
```diff
@@ -72,7 +72,10 @@ enum
   PROP_0,
```
Comment (mariospr) [inline src/st/st-entry.c]: Ok

File: js/ui/shellEntry.js [hunk 13]
```diff
@@ -121,6 +134,293 @@ const EntryMenu = new Lang.Class({
...
+        searchString = searchString.replace(/^\s+/g, '').replace(/\s+$/g, '');
+        if (searchString == '')
+            return [];
+
+        let terms = searchString.split(/\s+/);
+        return terms;
+    },
+
+    _onCapturedEvent: function(actor, event) {
+        if (event.type() == Clutter.EventType.BUTTON_PRESS) {
```
Comment (mariospr) [inline js/ui/shellEntry.js]: Ok

File: js/misc/util.js (hunk 2 above)
Comment (mariospr) [inline js/misc/util.js]: Ok

**Comment by mariospr** [comment]
@cosimoc I think I addressed all your comments with the exception of `addSettingsAction()` (see my comment about that), so this should be ready for another review, and maybe merging

File: js/ui/backgroundMenu.js (hunk 5 above)
Comment (cosimoc) [inline js/ui/backgroundMenu.js]: OK that makes sense, as long as we're using the speedwagon.

File: src/shell-grid-desaturate-effect.c [hunk 14]
```diff
@@ -39,8 +39,6 @@
 #define SHELL_IS_GRID_DESATURATE_EFFECT_CLASS(klass)     (G_TYPE_CHECK_CLASS_TYPE ((klass), SHELL_TYPE_GRID_DESATURATE_EFFECT))
 #define SHELL_GRID_DESATURATE_EFFECT_GET_CLASS(obj)      (G_TYPE_INSTANCE_GET_CLASS ((obj), SHELL_TYPE_GRID_DESATURATE_EFFECT, ShellGridDesaturateEffectClass))
 
-#define CLUTTER_ENABLE_EXPERIMENTAL_API
```
Comment (cosimoc) [inline src/shell-grid-desaturate-effect.c]: Please squash this with https://github.com/endlessm/gnome-shell/pull/11/commits/36ba8fdac4801f47861874497df38a8cf8fb8930

File: js/ui/search.js [hunk 15]
```diff
@@ -603,25 +604,51 @@ const SearchResults = new Lang.Class({
...
+    set isAnimating (v) {
+        if (this._isAnimating == v)
+            return;
+
+        this._isAnimating = v;
+        this._updateSearchProgress();
+        if (!this._isAnimating) {
+            this._providers.forEach(Lang.bind(this, function (provider) {
+                let results = this._results[provider.id];
+                if (results) {
```
Comment (cosimoc) [inline js/ui/search.js]: Remove curly braces for single-line if

File: js/ui/workspaceMonitor.js [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        this._shellwm.connect('minimize-completed', Lang.bind(this, this._updateOverview));
+        this._shellwm.connect('destroy-completed', Lang.bind(this, this._updateOverview));
+
+        this._metaScreen = global.screen;
+        this._metaScreen.connect('in-fullscreen-changed', Lang.bind(this, this._fullscreenChanged));
+
+        let primaryMonitor = Main.layoutManager.primaryMonitor;
+        this._inFullscreen = primaryMonitor && primaryMonitor.inFullscreen;
+
+        this._appSystem = Shell.AppSystem.get_default();
```
Comment (cosimoc) [inline js/ui/workspaceMonitor.js]: Instead of saving this into a private variable, just call the singleton getter in the only place where it's needed.

File: data/Makefile.am [hunk 17]
```diff
@@ -2,7 +2,7 @@ CLEANFILES =
 NULL =
 
 desktopdir=$(datadir)/applications
-desktop_DATA = org.gnome.Shell.desktop gnome-shell-extension-prefs.desktop eos-launch.desktop
+desktop_DATA = org.gnome.Shell.desktop gnome-shell-extension-prefs.desktop eos-launch.desktop eos-link-feedback.desktop
```
Comment (cosimoc) [inline data/Makefile.am]: If this only ends up being used by Yelp, I'd want to move it to that package.

File: src/st/st-entry.c [hunk 18]
```diff
@@ -241,10 +245,10 @@ st_entry_dispose (GObject *object)
   StEntryPrivate *priv = ST_ENTRY_PRIV (entry);
   GdkKeymap *keymap;
 
-  if (priv->entry)
```
Comment (cosimoc) [inline src/st/st-entry.c]: Sorry to nitpick, but I think this change should be separated out into another commit. It's not related to this IIRC -- see https://github.com/endlessm/eos-desktop/commit/b27e85340f197d82da5372a1f26baaff27e3f4d1

File: js/misc/util.js [hunk 19]
```diff
@@ -57,6 +61,29 @@ function findUrls(str) {
...
+//
+// Similar to "findUrls", but for use only with terms from the searchbar.
+// The regex for these URLs matches strings such as "google.com" (note the
+// lack of preceding scheme).
+//
+// Return value: the list of URLs found in the string
+function findSearchUrls(terms) {
+    let res = [], match;
+    for (let i in terms) {
+        while ((match = _searchUrlRegexp.exec(terms[i]))) {
```
Comment (cosimoc) [inline js/misc/util.js]: Remove curly braces for single-line condition

File: js/ui/internetSearch.js [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+// Returns a plain URI if the user types in
+// something like "facebook.com"
+function getURIForSearch(terms) {
+    let searchedUris = Util.findSearchUrls(terms);
+    // Make sure search contains only a uri
+    // Avoid cases like "what is github.com"
+    if (searchedUris.length == 1 && terms.length == 1) {
+        let uri = searchedUris[0];
+        // Ensure all uri has a scheme name
+        if (!GLib.uri_parse_scheme(uri)) {
```
Comment (cosimoc) [inline js/ui/internetSearch.js]: Remove these curlies too

File: js/ui/internetSearch.js [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+        }
+        return uri;
+    } else {
+        return null;
+    }
+}
+
+function getInternetSearchProvider() {
+    let browserApp = Util.getBrowserApp();
+    if (browserApp) {
```
Comment (cosimoc) [inline js/ui/internetSearch.js]: Remove these curlies too

File: js/ui/internetSearch.js [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+        return results.slice(0, maxNumber);
+    },
+
+    getInitialResultSet: function(terms, callback, cancellable) {
+        let results = [];
+
+        if (this._networkMonitor.network_available) {
+            let uri = getURIForSearch(terms);
+            let query = terms.join(' ');
+            if (uri) {
```
Comment (cosimoc) [inline js/ui/internetSearch.js]: Remove this and the below curlies too

File: js/ui/internetSearch.js [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+            this._launchURI(uri);
+        } else if (metaId.startsWith('search:')) {
+            let query = metaId.slice('search:'.length);
+            this._launchURI('? '.concat(query));
+        }
+    },
+
+    launchSearch: function(terms) {
+        this.getInitialResultSet(terms, Lang.bind(this, function(results) {
+            if (results) {
```
Comment (cosimoc) [inline js/ui/internetSearch.js]: Remove these curly braces

File: js/ui/search.js [hunk 24]
```diff
@@ -425,6 +426,11 @@ const SearchResults = new Lang.Class({
         this._cancellable = new Gio.Cancellable();
 
         this._registerProvider(new AppDisplay.AppSearchProvider());
+
+        this._internetProvider = InternetSearch.getInternetSearchProvider();
+        if (this._internetProvider) {
```
Comment (cosimoc) [inline js/ui/search.js]: Remove curly braces

File: src/shell-app-system.c [hunk 25]
```diff
@@ -207,14 +207,6 @@ app_is_stale (ShellApp *app)
   return !is_unchanged;
 }
 
-static gboolean
-stale_app_remove_func (gpointer key,
-                       gpointer value,
-                       gpointer user_data)
-{
-  return app_is_stale (value);
```
Comment (cosimoc) [inline src/shell-app-system.c]: Squash this commit into https://github.com/endlessm/gnome-shell/pull/11/commits/d7372ddaf8e7ee753f2a919c3af7d43afa1d7ce3

**Comment by cosimoc** [comment]
@mariospr Also please rebase the branch to master.

**Comment by mariospr** [comment]
> @mariospr Also please rebase the branch to master.

Sorry, it's rebased now

File: js/ui/search.js (hunk 24 above)
Comment (mariospr) [inline js/ui/search.js]: Ok

File: js/ui/search.js (hunk 15 above)
Comment (mariospr) [inline js/ui/search.js]: Ok

File: js/ui/workspaceMonitor.js (hunk 16 above)
Comment (mariospr) [inline js/ui/workspaceMonitor.js]: I'm not sure that's better (I kind of like calling the getter once only for the lifetime of workspaceMonitor), but ok

File: js/misc/util.js (hunk 19 above)
Comment (mariospr) [inline js/misc/util.js]: Ok

File: js/ui/internetSearch.js (hunk 20 above)
Comment (mariospr) [inline js/ui/internetSearch.js]: Ok

File: js/ui/internetSearch.js (hunk 21 above)
Comment (mariospr) [inline js/ui/internetSearch.js]: Ok

File: js/ui/internetSearch.js (hunk 22 above)
Comment (mariospr) [inline js/ui/internetSearch.js]: Ok

File: js/ui/internetSearch.js (hunk 23 above)
Comment (mariospr) [inline js/ui/internetSearch.js]: Ok

File: data/Makefile.am (hunk 17 above)
Comment (mariospr) [inline data/Makefile.am]: This commit is from @GeorgesStavracas, not sure how it ended up in this PR (should be gone now)

In any case, I agree with you that this should be moved to Yelp

File: src/st/st-entry.c (hunk 18 above)
Comment (mariospr) [inline src/st/st-entry.c]: Done
